### ISSUE TITLE
Unexpected error in control flow loop analysis

### ISSUE THREAD

**Issue by ahejlsberg** [Microsoft/TypeScript#11000]
The following two functions are functionally equivalent, but the type checker reports an unexpected error in the second one:

```ts
declare var cond: boolean;

function foo1() {
    let x: string | number | boolean = 0;
    while (cond) {
        if (typeof x === "string") {
            x = x.slice();
        }
        else {
            x = "abc";
        }
    }
}

function foo2() {
    let x: string | number | boolean = 0;
    while (cond) {
        if (typeof x === "number") {
            x = "abc";
        }
        else {
            x = x.slice();  // Error: Property 'slice' does not exist on type 'string | boolean'
        }
    }
}
```

This issue is related to our handling of incomplete types during control flow analysis of loops. When a type guard for an incomplete type produces `never` we revert to the *declared* type of the variable. That causes `boolean` to spuriously appear in the second example above.
